### DISCUSSION TITLE
What replaced DefaultInputSlot in the ipp world?

### DISCUSSION THREAD

**Discussion by RolandHughes** [OpenPrinting/cups-filters#607] [Q&A]
All,

Trying to upgrade some legacy PPD parsing library code to the new ipp stuff. This has me stumped. Function to determine defaultInputSlot()

`    // NOTE: Implemented in both CUPS and Mac plugins, please keep in sync
    // Try load standard PPD option first
    if ( m_ppd )
    {
        ppd_option_t *inputSlot = ppdFindOption( m_ppd, "DefaultInputSlot" );

        if ( inputSlot )
        {
            return QPrintUtils::ppdChoiceToInputSlot( inputSlot->choices[0] );
        }

        // If no result, then try a marked option
        ppd_choice_t *defaultChoice = ppdFindMarkedChoice( m_ppd, "InputSlot" );

        if ( defaultChoice )
        {
            return QPrintUtils::ppdChoiceToInputSlot( *defaultChoice );
        }
    }

    // Otherwise return Auto
    return QPlatformPrintDevice::defaultInputSlot();
`
I can't find any documentation on what this is called in the "new" reality. Would hate to have to find some way to make Cups generate an in-memory PPD just to still support this.

Dumping attributes for all the printers on my network I see this:
`media-col-default={media-type=stationery media-size={x-dimension=21590 y-dimension=27940} media-bottom-margin=432 media-left-margin=432 media-right-margin=432 media-top-margin=432 media-source=auto media-auto-dimension=false media-source-properties={media-source-feed-direction=short-edge-first media-source-feed-orientation=3}}
`
Do I need to pull this attribute then parse the string for "media-source=" ?
On a miss, like would happen for my HP OfficeJet

`media-col-default={media-size={x-dimension=21590 y-dimension=27940} media-top-margin=330 media-bottom-margin=330 media-left-margin=330 media-right-margin=330 media-source=main media-type=stationery}
media-source-supported=main,alternate
`
then just use the first entry in media-source-supported?

Really would like to know what is supposed to be done here.

Is this whole concept of "input slot" being purged from the Cups universe? I could understand that since most of the new-ish printers

`media-source-supported=auto,by-pass-tray,tray-1`

would almost always have their default set to "auto."

Please advise.

Thanks,

**Comment by zdohnal** [comment]
Hi,

based on the [IPP specification](https://ftp.pwg.org/pub/pwg/candidates/cs-ippjobext21-20230210-5100.7.pdf) media-col-default is the right place for media related defaults, but you don't need to parse the string by yourself, but use CUPS API. Taken from OpenPrinting CUPS cups/ppd-cache:

```
3583   if ((defattr = ippFindAttribute(supported, "media-col-default", IPP_TAG_BEGIN_COLLECTION)) != NULL)
3584   {
...
3922  if ((attr = ippFindAttribute(ippGetCollection(defattr, 0), "media-source", IPP_TAG_ZERO)) != NULL)
```

That way you get media-source attribute, which contains the default value regarding the input tray.
